docs/site/Parsing-requests.md

```diff
@@ -0,0 +1,159 @@
+tags:
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Parsing-requests.html
+summary:
+---
+
+## Parsing Requests
+
+This is an action in the default HTTP sequence, it parses arguments from an
+incoming request and uses them as inputs to invoke the corresponding controller
+method.
+
+This action contains 3 steps:
+
+- Parses arguments from request query, body, path and header according to the
+  operation's OpenAPI specification.
+- Coerces parameters from string to its corresponding JavaScript run-time type.
+- Performs validation on the parameters and body data.
+
+### Parsing Raw Data
+
+The code below defines a typical endpoint by decorating a controller method with
+rest decorators.
+
+```ts
+class TodoController() {
+  @put('/todos/{id}')
+  async replaceTodo(
+    @param.path.number('id') id: number,
+    @requestBody() todo: Todo,
+  ): Promise<boolean> {
+    return await this.todoRepo.replaceById(id, todo);
+  }
+}
+```
+
+An operation specification will be generated in-memory to describe it, and raw
+datas are parsed from request according to the specification. For example the
+first parameter is from source `path`, then its value will be parsed from a
+request's path.
+
+_See [controller](Controller.md) for more details of defining an endpoint._
+
+_See
+[OpenAPI operation object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#operationObject)
+to know more about its structure._
+
+### Coercion
+
+The parameters parsed from path, header, query of a http request are always in
+the string format. Therefore when invoking a controller function, a parameter
+need to be converted to its corresponding JavaScript runtime type, which is
+inferred from its parameter specification.
+
+Give the example operation `replaceTodo` in section
+[parsing raw data](#parsing-raw-data), which takes in a number `id` as the first
+input, coercion means users don't need to do a type conversion by themselves in
+the controller function like below:
+
+```ts
+@put('/todos/{id}')
+async replaceTodo(
+  @param.path.number('id') id: number,
+  @requestBody() todo: Todo,
+): Promise<boolean> {
+  // NO need to do the "string to number" convertion as below
+  id = +id;
+  return await this.todoRepo.replaceById(id, todo);
+}
+```
+
+### Validation
+
+Validations are applied on the parameters and the request body data. They also
+use OpenAPI specification as the reference to infer the validation rules.
+
+#### Parameters
+
+We have the data type safety check for the parameters parsed from header, path,
+query. For example, if a parameter should be an integer, then a number with
+decimal like "1.23" will be rejected.
+
+You can specify a parameter's type by calling shortcut decorators of `@param`
+like `@param.query.integer()`. A list of available shortcuts can be found in the
+[api documents](https://apidocs.strongloop.com/@loopback%2fdocs/openapi-v3.html#param).
+And please check [parameter decorators](Decorators.md#parameter-decorator) for
```

Comment:
`Check out the section on parameter decorators(link) for instructions on how to decorate the ...`